File: userPI/src/main/java/org/example/userpi/service/UserService.java
package org.example.userpi.service;

import org.example.userpi.dto.ChangePasswordRequest;
import org.example.userpi.dto.UserDirectoryDto;
import org.example.userpi.model.User;
import java.util.List;

public interface UserService {
    List<User> getAllUsers();

    List<UserDirectoryDto> getDirectoryUsers();
    User getUserById(int id);
    User updateUser(User user, int requestingUserId);
    User addUser(User user);
    void deleteUser(int id);
    void setPassword(int id, String password, int requestingUserId);
    void changePassword(int id, ChangePasswordRequest request, int requestingUserId);
}

File: PIcloud/src/app/core/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';
import { User } from '../models/user.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {

  private apiUrl = 'http://localhost:8090/api/users';

  constructor(private http: HttpClient) {}

  async getAllUsers(): Promise<User[]> {
    return firstValueFrom(this.http.get<User[]>(this.apiUrl));
  }

  async getUserById(id: number): Promise<User> {
    return firstValueFrom(this.http.get<User>(`${this.apiUrl}/${id}`));
  }

  async updateUser(user: User): Promise<User> {
    return firstValueFrom(this.http.put<User>(this.apiUrl, user));
  }

  async deleteUser(id: number): Promise<void> {
    return firstValueFrom(this.http.delete<void>(`${this.apiUrl}/${id}`));
  }
  
  async setPassword(id: number, password: string): Promise<void> {
  return firstValueFrom(
    this.http.post<void>(`${this.apiUrl}/${id}/set-password`, { password }, {
      responseType: 'text' as 'json'
    })
  );
}

  async changePassword(id: number, oldPassword: string, newPassword: string): Promise<void> {
    return firstValueFrom(
      this.http.put<void>(`${this.apiUrl}/${id}/change-password`, {
        oldPassword,
        newPassword
      })
    );
  }
}